### ISSUE TITLE
Re-enabling native compilation in Emacs

### ISSUE THREAD

**Issue by svraka** [msys2/MINGW-packages#17374]
Recently native compilation support was turned off in bfa4238 while merging https://github.com/msys2/MINGW-packages/pull/17129 due to build timeouts. The reason given in the commit message was slow compilation under GCC 13. Based on my testing I don't think that's the case, although the build still fails in https://github.com/svraka/MINGW-packages/tree/emacs-libgccjit.

After reverting bfa4238 the CI gets stuck ([CI summary](https://github.com/svraka/MINGW-packages/actions/runs/5115456102/jobs/9196734675)) at exactly the same step as a couple of weeks ago with @Biswa96's commit ([CI summary](https://github.com/msys2/MINGW-packages/actions/runs/4967459226/jobs/8889468717?pr=17129)), with the log displaying `ELN  ../../emacs-28.2/lisp/x-dnd.eln`.

The build reached this point fairly quickly, in just over 10 minutes in both cases. In my local testing the whole build from a fully up-to-date MSYS2 takes about 10 minutes (the CI would take more, as it needs to install dependencies) with `-j2` (emulating the GitHub runner with two cores).

Given how `make` handles targets in parallel builds, this no guarantee that we are stuck exactly at the same point but it is at least *very* suspicious.

After adding [Telemetry](https://github.com/marketplace/actions/workflow-telemetry) to the CI actions (see [CI summary](https://github.com/svraka/MINGW-packages/actions/runs/5121360287)) I don't see any resource bottlenecks but the CPU graph is pretty strange. CPU usage in the first 20 minutes what I would expect. But later it looks like there's not much going on.

So I added the [debugging with tmate](https://github.com/marketplace/actions/debugging-with-tmate) action to the CI steps and SSHd into the CI machine. `lisp/x-dnd.eln` in the source tree is successfully native compiled, and at this point native compilation of `lisp/files.eln` is the only running target but it's hung with 0% CPU usage (linebreaks added for clarity):

```
C:\_\B\src\build-UCRT64\src\emacs.exe \
    -batch --no-site-file --no-site-lisp \
    --eval "(setq load-prefer-newer t)" \
    -l comp -f byte-compile-refresh-preloaded \
    --eval "(batch-native-compile t)" ../../emacs-28.2/lisp/files.el
```

I can kill this process and the build will successfully finish without native compiling `files.el`.

Can someone try building Emacs with native compilation support on their machine? I have no idea what could be the problem with the CI.

**Comment by lb90** [comment]
Thanks, that's very helpful! Could you also post the stacktrace to https://debbugs.gnu.org/cgi/bugreport.cgi?bug=63365?

**Comment by oscarfv** [comment]
The backtrace indicates that crash happens and then the process gets stuck with a modal dialog, possibly the dialog that reports the crash.

**Comment by cyrilarnould** [comment]
> Can someone try building Emacs with native compilation support on their machine? I have no idea what could be the problem with the CI.

Weirdly enough, when I build `mingw-w64-emacs` from the `MINGW-packages` repository on my local machine, the build finishes with no issues. I use the following command, is this any different from what's happening in the CI environment?

```
MINGW_ARCH=mingw64 makepkg-mingw -sCLf
```

On the other hand, if I check out the `emacs` repository and build from there (with all default options), the build gets stuck as well, but in a different spot (`ELC textmodes/sgml-mode.elc`).The steps to reproduce:

```
git clone https://git.savannah.gnu.org/git/emacs.git
cd emacs
git checkout emacs-28.2
./autogen.sh
./configure --with-native-compilation
make
```

This uses  `-g3 -O2 -gdwarf-2` as `CFLAGS`. Building using `makepkg-mingw` uses `-march=nocona -msahf -mtune=generic -O2 -pipe -fstack-protector-strong`. Maybe we can get the CI build to work by playing around with other options? In #17319, using `-O0` was working.

**Comment by oscarfv** [comment]
> Maybe we can get the CI build to work by playing around with other options?

`-fstack-protector-strong` looks promising.

**Comment by svraka** [comment]
> Thanks, that's very helpful! Could you also post the stacktrace to https://debbugs.gnu.org/cgi/bugreport.cgi?bug=63365?

Done.

Also did a quick hack to downgrade GCC and seems to work: https://github.com/svraka/MINGW-packages/actions/runs/5141382943

**Comment by svraka** [comment]
> > Maybe we can get the CI build to work by playing around with other options?
> 
> `-fstack-protector-strong` looks promising.

Nope, still the same.

https://github.com/svraka/MINGW-packages/actions/runs/5146388820/jobs/9265358081#step:13:2012

**Comment by lb90** [comment]
Well done, @svraka! We could bisect the commit in GCC that introduced the regression, but it's a lot of work :/

**Comment by svraka** [comment]
Fixed in #17549.
